apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,423 @@
+    It = It0#it{last_seen_key = Key},
+    case emqx_ds_bitmask_keymapper:bin_checkmask(Filter, Key) of
+        true ->
+            Msg = deserialize(Val),
+            case check_message(Cutoff, It, Msg) of
+                true ->
+                    Acc = [Msg | Acc0],
+                    traverse_interval(ITHandle, Filter, Cutoff, It, Acc, N - 1);
+                false ->
+                    traverse_interval(ITHandle, Filter, Cutoff, It, Acc0, N);
+                overflow ->
+                    {0, It0, Acc0}
+            end;
+        false ->
+            {N, It, Acc0}
+    end.
+
+traverse_interval(_ITHandle, _Filter, _Cutoff, It, Acc, 0) ->
+    {0, It, Acc};
+traverse_interval(ITHandle, Filter, Cutoff, It, Acc, N) ->
+    inc_counter(),
+    case rocksdb:iterator_move(ITHandle, next) of
+        {ok, Key, Val} ->
+            traverse_interval(ITHandle, Filter, Cutoff, Key, Val, It, Acc, N);
+        {error, invalid_iterator} ->
+            {0, It, Acc}
+    end.
+
+-spec check_message(emqx_ds:time(), iterator(), emqx_types:message()) ->
+    true | false | overflow.
+check_message(
+    Cutoff,
+    _It,
+    #message{timestamp = Timestamp}
+) when Timestamp >= Cutoff ->
+    %% We hit the current epoch, we can't continue iterating over it yet.
+    %% It would be unsafe otherwise: messages can be stored in the current epoch
+    %% concurrently with iterating over it. They can end up earlier (in the iteration
+    %% order) due to the nature of keymapping, potentially causing us to miss them.
+    overflow;
+check_message(
+    _Cutoff,
+    #it{start_time = StartTime, topic_filter = TopicFilter},
+    #message{timestamp = Timestamp, topic = Topic}
+) when Timestamp >= StartTime ->
+    emqx_topic:match(emqx_topic:words(Topic), TopicFilter);
+check_message(_Cutoff, _It, _Msg) ->
+    false.
+
+format_key(KeyMapper, Key) ->
+    Vec = [integer_to_list(I, 16) || I <- emqx_ds_bitmask_keymapper:key_to_vector(KeyMapper, Key)],
+    lists:flatten(io_lib:format("~.16B (~s)", [Key, string:join(Vec, ",")])).
+
+format_keyfilter(any) ->
+    any;
+format_keyfilter({Op, Val}) ->
+    {Op, integer_to_list(Val, 16)}.
+
+-spec make_key(s(), emqx_types:message()) -> {binary(), [binary()]}.
+make_key(#s{keymappers = KeyMappers, trie = Trie}, #message{timestamp = Timestamp, topic = TopicBin}) ->
+    Tokens = emqx_topic:tokens(TopicBin),
+    {TopicIndex, Varying} = emqx_ds_lts:topic_key(Trie, fun threshold_fun/1, Tokens),
+    VaryingHashes = [hash_topic_level(I) || I <- Varying],
+    KeyMapper = array:get(length(Varying), KeyMappers),
+    KeyBin = make_key(KeyMapper, TopicIndex, Timestamp, VaryingHashes),
+    {KeyBin, Varying}.
+
+-spec make_key(emqx_ds_bitmask_keymapper:keymapper(), emqx_ds_lts:static_key(), emqx_ds:time(), [
+    non_neg_integer()
+]) ->
+    binary().
+make_key(KeyMapper, TopicIndex, Timestamp, Varying) ->
+    UniqueInteger = erlang:unique_integer([monotonic, positive]),
```

Comment:
Q: why do we add this unique integer to the key?

Reply:
This is to avoid losing messages published to the same topic in a very short span of time (within the same millisecond). RocksDB would overwrite the old record in this case.